Library/OcPciIoLib/OcPciIoU.c

```diff
@@ -0,0 +1,446 @@
+   be handled by the driver.
+
+   If Width is EfiCpuIoWidthUint8, EfiCpuIoWidthUint16, EfiCpuIoWidthUint32,
+   or EfiCpuIoWidthUint64, then both Address and Buffer are incremented for
+   each of the Count operations that is performed.
+
+   If Width is EfiCpuIoWidthFifoUint8, EfiCpuIoWidthFifoUint16,
+   EfiCpuIoWidthFifoUint32, or EfiCpuIoWidthFifoUint64, then only Buffer is
+   incremented for each of the Count operations that is performed. The read or
+   write operation is performed Count times on the same Address.
+
+   If Width is EfiCpuIoWidthFillUint8, EfiCpuIoWidthFillUint16,
+   EfiCpuIoWidthFillUint32, or EfiCpuIoWidthFillUint64, then only Address is
+   incremented for each of the Count operations that is performed. The read or
+   write operation is performed Count times from the first element of Buffer.
+
+   @param[in]  This     A pointer to the EFI_CPU_IO2_PROTOCOL instance.
+   @param[in]  Width    Signifies the width of the I/O or Memory operation.
+   @param[in]  Address  The base address of the I/O operation.
+   @param[in]  Count    The number of I/O operations to perform. The number of
+                       bytes moved is Width size * Count, starting at Address.
+   @param[out] Buffer   For read operations, the destination buffer to store the results.
+                       For write operations, the source buffer from which to write data.
+
+   @retval EFI_SUCCESS            The data was read from or written to the PI system.
+   @retval EFI_INVALID_PARAMETER  Width is invalid for this PI system.
+   @retval EFI_INVALID_PARAMETER  Buffer is NULL.
+   @retval EFI_UNSUPPORTED        The Buffer is not aligned for the given Width.
+   @retval EFI_UNSUPPORTED        The address range specified by Address, Width,
+                                 and Count is not valid for this PI system.
+
+ **/
+EFI_STATUS
+EFIAPI
+CpuMemoryServiceRead (
+  IN EFI_CPU_IO2_PROTOCOL       *This,
+  IN EFI_CPU_IO_PROTOCOL_WIDTH  Width,
+  IN UINT64                     Address,
+  IN UINTN                      Count,
+  OUT VOID                      *Buffer
+  )
+{
+  EFI_STATUS                 Status;
+  UINT8                      InStride;
+  UINT8                      OutStride;
+  EFI_CPU_IO_PROTOCOL_WIDTH  OperationWidth;
+  UINT8                      *Uint8Buffer;
+
+  Status = CpuIoCheckParameter (TRUE, Width, Address, Count, Buffer);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  //
+  // Select loop based on the width of the transfer
+  //
+  InStride       = mInStride[Width];
+  OutStride      = mOutStride[Width];
+  OperationWidth = (EFI_CPU_IO_PROTOCOL_WIDTH)(Width & 0x03);
+  for (Uint8Buffer = Buffer; Count > 0; Address += InStride, Uint8Buffer += OutStride, Count--) {
+    if (OperationWidth == EfiCpuIoWidthUint8) {
+      *Uint8Buffer = MmioRead8 ((UINTN)Address);
+    } else if (OperationWidth == EfiCpuIoWidthUint16) {
+      *((UINT16 *)Uint8Buffer) = MmioRead16 ((UINTN)Address);
```

Comment:
Use WriteUnalignedXX functions here and below.

Reply:
There doesn't seem to be one for Uint8 though, same for below

Reply:
Well, obviously, since any area is 1-byte aligned. Only 16+ are affected by undefined behaviour in the current code.

Reply:
Is this needed ? I can't figure this out without causing a compiler warning under GCC and MSVC. Using `WriteUnaligned32 ((UINT32 *)Address, *((UINT32 *)Uint8Buffer));` works fine under Clang/Xcode but gives a compiler warning on MSVC / GCC, so I tried using bitmask which usually fixes this error. 

`WriteUnaligned32 ((UINT32 *)(Address & 0xFFFFFFFF), *((UINT32 *)Uint8Buffer));` causes function to not work (freeze on connecting drivers). Tried `0xFFFFFFFFULL` too.

Maybe I'm missing something.

Reply:
Yes, this is needed. `(UINT32 *)(UINTN)Address`. Truncating the address on 64-bit is borked for obvious reasons.

Reply:
That worked, thanks!